pkg/sql/schemachanger/scbuild/internal/scbuildstmt/alter_table_alter_primary_key.go

```diff
@@ -0,0 +1,384 @@
+// Copyright 2022 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package scbuildstmt
+
+import (
+	"fmt"
+
+	"github.com/cockroachdb/cockroach/pkg/clusterversion"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/catpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/colinfo"
+	"github.com/cockroachdb/cockroach/pkg/sql/catalog/tabledesc"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgcode"
+	"github.com/cockroachdb/cockroach/pkg/sql/pgwire/pgerror"
+	"github.com/cockroachdb/cockroach/pkg/sql/privilege"
+	"github.com/cockroachdb/cockroach/pkg/sql/schemachanger/scpb"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/catid"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/util/errorutil/unimplemented"
+	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
+	"github.com/cockroachdb/errors"
+)
+
+// The implementation of `ALTER PRIMARY KEY` will be broken into multiple PRs:
+// 1. Implement the simplest case where we drop the old primary index and add the new primary index;
+// 2. Consider the case where the new primary key is requested to be sharded.
+// 3. Consider the case where the old primary index is on the implicitly created `rowid` column,
+//    in which case we also need to drop that column;
+// 4. Consider the case where altering primary key requires us to modify existing secondary indexes
+//    (see the legacy schema change about in what cases we should rewrite)
+// 5. Consider partitioning and locality (I'm not sure what they are, and why they play a role when
+//    `ALTER PRIMARY KEY` but I've seen them in the old schema changer, so I assume we ought to do
+//    something about them too here).
+func alterTableAlterPrimaryKey(
+	b BuildCtx, tn *tree.TableName, tbl *scpb.Table, t *tree.AlterTableAlterPrimaryKey,
+) {
+	// Panic on certain forbidden `ALTER PRIMARY KEY` cases (e.g. one of
+	// the new primary key column is a virtual column). See the comments
+	// for a full list of preconditions we check.
+	checkForEarlyExit(b, tbl, t)
+
+	// TODO (xiang): This section contains all fall-back cases and need to
+	// 							 be removed to fully support `ALTER PRIMARY KEY`.
+	fallBackIfRequestToBeSharded(t)
+	fallBackIfRegionalByRow(b, tbl.TableID)
+
+	tableElems := b.QueryByID(tbl.TableID)
+
+	// Retrieve old primary index and its name elements.
+	oldPrimaryIndexElem := mustRetrievePrimaryIndexElement(tableElems, tn.String())
+	oldPrimaryIndexNameElem := mustRetrievePrimaryIndexNameElem(tableElems, oldPrimaryIndexElem.IndexID, tn.String())
+
+	// Resolve and drop elements from the old primary index
```

Comment:
How does this compose with adding or dropping columns? Let's look at this together tomorrow afternoon. In the meantime, can I encourage you to read the DROP COLUMN PR